packages/combobox/src/index.tsx

```diff
@@ -373,6 +376,40 @@ if (__DEV__) {
   };
 }
 
+/**
+ * Uses a ref object which stores the index as a key and custom data as value
+ * for each ComboboxOption. Hides the ref so that we can only mutate it through
+ * the returned functions. 🙈
+ */
+const useOptionDataFactory = (): {
```

Comment:
@CodingDive I can see the value in having the option index, but I'm curious what your thoughts on are on the use of the `data` object prop.

Reply:
Arbitrary `data` is a lot more flexible than just `index`.
I might want to pass the whole item, or just the id of an item when using a record (e.g `Record<id, Item>`).
Index is mostly relevant for arrays and even then, I wouldn't feel great to use the index of an selected item, knowing that the list we use for rendering can change at any point (filtering/sorting, etc.). As soon as the selected index is stored in state and the list is changed, the consecutive rerenders after `onSelect` might be dealing with a completely wrong or non existing item.

Reply:
Apologies, I meant id rather than index. I'm still not sure I understand the full value of passing in any amount of arbitrary data (when that data could be accessed outside of the Combobox if it were keyed by an id).

Perhaps it would be good for us to show a problem that this addition would solve. Can you think of a good example case other than or in addition to the [example that alerts the selected item's index][1]?

[1]: https://github.com/reach/reach-ui/pull/514/files#diff-fcbf2f9176bec0e87f86974e979bb125R30

Reply:
I think, it just makes working with onSelect easier. You just put in some object as option value, and you receive exact same object in onSelect. No need to figure out how to first get selected id and then how to get option by id.

Check out my suggestion please - https://github.com/reach/reach-ui/pull/514#issuecomment-609458433

Reply:
The main use case would be to have a really big array and not have to `.find` the selected item once again. Changed the example and prop docs to reflect that other things as id or index can be passed in.